### ISSUE TITLE
Support proposed ES Next "::" bind operator

### ISSUE THREAD

**Issue by loucyx** [microsoft/TypeScript#3508]
*Edit from @DanielRosenwasser: Since this issue was filed, a separate proposal for `::` was created called "extensions".*

https://github.com/tc39/proposal-extensions
___

It will be great if TypeScript had an implementation of the "::" bind operator of ECMAScript 7. Some examples of this operator and the implementation in ES5:
**Bind**

``` typescript
// ES5
var log = console.log.bind(console);

// ES7
let log = ::console.log;
```

**Call**

``` typescript
// ES5
var forEach = Array.prototype.forEach,
    elements = document.querySelectorAll("div");

forEach.call(elements, div => { console.log(div); })

// ES7
let forEach = Array.prototype.forEach,
    elements = document.querySelectorAll("div");

elements::forEach(div => { console.log(div) });
```

[[ref](http://wiki.ecmascript.org/doku.php?id=strawman:bind_operator)] [[more examples](http://blog.jeremyfairbank.com/javascript/javascript-es7-function-bind-syntax/)]

**Comment by DanielRosenwasser** [comment]
I keep seeing this being called an ES7 operator. The strawman proposal has been up on the wiki for quite a while, but [I don't see any proposal for it right now](https://github.com/tc39/ecma262/blob/c87d1f5b65067865bab3e1ac81b6d35e2553b515/README.md). I've tentatively marked it as ES7 anyway, but perhaps someone else can weigh in on this.

**Comment by kitsonk** [comment]
They implemented it in [Babel](http://babeljs.io/blog/2015/05/14/function-bind/) but even they say it is highly experimental.  While it might be a good idea, it is far away from being a standard and no runtime has implemented it.

**Comment by sebmck** [comment]
@DanielRosenwasser It was discussed at the March TC39 meeting (notes [here](https://github.com/tc39/tc39-notes/blob/master/es6/2015-03/mar-25.md#6vi-function-bind-and-private-fields-redux-kevin-smith)). It's also an important distinction to draw that Babel and TypeScript have completely different priorities and features implemented in one are not necessarily appropriate for both.

**Comment by ivogabe** [comment]
I think this is the best way to implement extension methods (#9), since with this proposal the emit isn't based on type info.

Also this will (partially) solve #212 (bind, call & apply are untyped), since we can use the `::` operator to bind methods:
```typescript
// Bind
func.bind(obj);
obj::func;

// Call
func.call(obj, 42);
obj::func(42);

// Apply
func.apply(obj, arr);
obj::func(...arr);
```
We don't need bind, call and apply any more if we use the bind operator. The only missing feature of the bind operator would be currying, but I'm not sure whether that's used a lot. Personally I'm not using it.

Besides that, maybe we can emit a function expression instead of a bind call. Using bind() isn't good for performance and it isn't supported in ES3. I've written some code that could be generated. I've added temporary variables as the expression (`console.log` in this example) should be evaluated once I guess. My vote would be for the function expression / lambda approach.
```javascript
// Source
foo = ::console.log;

// .bind (ES5, 6)
foo = (_a = console, _a.log.bind(_a));
var _a, _b;

// lambda (ES6)
foo = (_a = console, _b = _a.log, (...args) => _b.apply(_a, args));
var _a, _b;

// Function expression (ES3, 5)
foo = (_a = console, _b = _a.log, function() { _b.apply(_a, arguments); });
var _a, _b;
```
About function expression, I'm not sure whether .apply() accepts an ArrayLike, so we might need to wrap arguments inside Array.prototype.slice.call() to convert it to an array.

Since this feature is not standardized yet, we could add it first under an `--experimentalBind` flag.

**Comment by HerringtonDarkholme** [comment]
I personally think supporting a new feature without type safety is desirable. FYI, https://github.com/Microsoft/TypeScript/issues/1985

**Comment by filipstachura** [comment]
Here is more detailed description of the bind operator strawman:
https://github.com/zenparsing/es-function-bind

I must say that syntax proposed by Trine (https://github.com/jussi-kalliokoski/trine) looks way better than anything I saw before. I would love to write code in such functional form.

**Comment by rbuckton** [comment]
This feature would work well with an ability to specify the type of `this` in a function declaration (#229).

**Comment by Artazor** [comment]
@rbuckton are there any chances to parse the following syntax: https://github.com/Microsoft/TypeScript/issues/1985#issuecomment-127293978 ?

**Comment by RyanCavanaugh** [comment]
Our policy on ES7 features is to wait until they're in Stage 1. Someone should chime in here when that happens for this feature.

**Comment by lucasmciruzzi** [comment]
Great @RyanCavanaugh ! ... I'll try to check daily until is on Stage 1 :)

**Comment by mhegazy** [comment]
related to #3694

**Comment by basarat** [comment]
Note : @rbuckton has done some work on this : https://github.com/Microsoft/TypeScript/commit/8521632d1c2ee34939de8dbde6915c6d259e0c1d (mentioning it for community knowledge :rose:)

**Comment by donaldpipowitch** [comment]
Looking forward to this. I hope this will become the de facto standard how future libraries like Lodash will work so we can do this `[ 1, 2, 3, 4].map(i => i * i)::chunk(2);` instead of something like `_.chunk` or `Array.prototype.chunk`.

**Comment by rbuckton** [comment]
The problem with `::` is that there's disagreement on how it should work. One camp believes that the left-hand side of the `::` operand should be provided as the `this` value to the function on the right-hand side. Another camp believes it should be provided as the first argument, to be compatible with existing libraries like lodash, etc. I chose to implement `::` using the same semantics as Babel; however, we could both be wrong by the time this feature reaches Stage 2.

**Comment by basarat** [comment]
Note that the implementation done is : provided as the first argument. Of course lets wait for the spec to finalize.

**Comment by donaldpipowitch** [comment]
My number one use case for `::` are future forward polyfills which _can_ change and therefor shouldn't pollute `prototype`. E.g. I'd like to go to MDN and take any polyfill (like [`map`](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/map#Polyfill) and it should behave in the same way with `::` or without (say `[].map === []::map`).

Like:
1. We want to polyfill an unstable spec → use `foo::polyfilledFunctionBar`.
2. We want to polyfill a stable spec → use `foo.polyfilledFunctionBar`.
3. Switch to the native implementation, if available → use `foo.nativeFunctionBar`.

**Comment by jods4** [comment]
@basarat 
> Note that the implementation done is : provided as the first argument.

but @rbuckton  said
> I chose to implement `::` using the same semantics as Babel

Babel is (currently) providing the left-hand side as `this` so I find basarat's precision rather confusing.
Looking at the code it seems to me that TS currently binds to `this`.

**Comment by basarat** [comment]
> Babel is (currently) providing the left-hand side as this so I find basarat's precision rather confusing.
> Looking at the code it seems to me that TS currently binds to this.

@jods4 You are right :  [TypeScript](https://github.com/Microsoft/TypeScript/compare/bindOperator#diff-5168a4e0635b0e71dad0e00bce8c066bR27) and also [the babel version](https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=let%20z%20%3D%20a%3A%3Ab%3B) both work by binding `this`.

Don't know why I read it wrong. Sorry :rose:

**Comment by sebmck** [comment]
Just to clear things up. The semantics are not "Babel" or "TypeScript". They're what are included in the current proposal.

**Comment by prabirshrestha** [comment]
What about supporting this feature under an experimental flag similar to `experimentalDecorators`? Would really love to start using this feature in TS.

In order to support either `this` or the first arg a flag similar to `--jsx react/preserve` could be used.
`--experimentalBind arg/this`.

**Comment by tejacques** [comment]
I'd love to have this in under an experimental flag. If the TypeScript team would be open to that I think it's very likely the community would assist with the feature.

I know that this feature would be incredibly useful to [RxJs](https://github.com/ReactiveX/RxJS/issues/541) as well as providing an extremely nice and intuitive way of dealing with extension methods.

**Comment by kitsonk** [comment]
I think the desire to keep things that haven't progressed far enough in the TC39 stages is wise, because you never know when [the rug](https://esdiscuss.org/topic/an-update-on-object-observe) will get pulled out from you.

**Comment by jeffbcross** [comment]
We could really use this feature on Angular and RxJS-Next. More so for users' benefit than for the frameworks themselves, so we can encourage users to use Observable operators as functions instead of having a large Observable prototype.

**Comment by kitsonk** [comment]
Best to rally the troops in TC39 then...

**Comment by d180cf** [comment]
The bind operator looks like an attempt to get partial classes without the need to modify prototypes. This looks great as long as these methods-extensions have unique names. But it seems to me that many such methods will have similar names, e.g. `each` and `forEach`, and there will be conflicts between different libraries. To resolve the name conflicts, we will have to introduce namespaces and the code won't look that nice anymore. I'm taking an example from github.com/zenparsing/es-function-bind:

```js
// Create bindings for just the methods that we need
let { find, html } = jake;

// Find all the divs with class="myClass", then get all of the "p"s and
// replace their content.
document.querySelectorAll("div.myClass")::find("p")::html("hahaha");
```

Obviously, 5 lines later in the code we want to use a lodash-like library to find an element in an array:

```js
let { find, flatten } = lodash;
items::flatten()::find(x => x % 3 == 0);
```

But the `find` variable is already declared! The workaround is simple, but kind of defeats the purpose of the bind operator:

```js
items::lodash.flatten()::lodash.find(x => x % 3 == 0);
```

As you see, what we're essentially doing here is invoking functions in the postfix style, so the bind operator is just a postfix-call operator. This is still a great thing to have. I don't see a reason, though, why not to introduce the prefix-call operator that doesn't require parens (and it seems such an operator has already been introduced: `@throttled(1500) @debounced @async x => new Promise(...)`, but its scope has been artificially limited).

**Comment by weswigham** [comment]
@d180cf just use modules.

**Comment by prabirshrestha** [comment]
@d180cf Here is an example on how conflicts could be solved.

```
import { find as $find, html } from 'jake';
import { find, flatten } from 'lodash';

document.querySelectorAll("div.myClass")::$find("p")::html("hahaha");
items::flatten()::find(x => x % 3 == 0);
```

**Comment by jods4** [comment]
@d180cf 
TS is going to implement what TC39 standardizes, so if you have concerns about the feature you should probably raise your voice there.

BTW like the two comments before me, I don't quite see how this problem is unique to the bind operator. 

Importing functions with identical name is not allowed even in "normal" code and the solutions are the same. If I faced such a situation I would do like @prabirshrestha, rename them `$find` and `_find` probably.

There are plenty of use-cases that are unlikely to clash where this operator is still useful. For example passing an instance method to an event handler: `$(a).on("click", ::this.clickHandler)`.

**Comment by igl** [comment]
Nicer: https://github.com/mindeavor/es-pipeline-operator

**Comment by niieani** [comment]
@igl Pipeline operator is pretty nice, but it serves a different purpose. I'd love to see them both implemented in TS.

**Comment by ivan-kleshnin** [comment]
Pipeline operators (`|>`, `<|`...) are for chaining functions. Bind operator `::` is for solving expression problem for object-oriented programming. Do not confuse them, please. As most JS has object-oriented API now the second one is much more important. And for func chaining you already have `Ramda.compose` / `Ramda.pipe` / etc.

**Comment by irakliy81** [comment]
+1

**Comment by aluanhaddad** [comment]
@ivan-kleshnin 
>And for func chaining you already have Ramda.compose / Ramda.pipe / etc

and for `::` we already have `Function.prototype.bind`.

Also, I fail to see how a syntactic sugar for `Function.prototype.bind`, as much as it would increase readability, has anything to do with solving the expression problem. It does not provide any new extensible mechanisms but rather simply makes invoking a function read more like invoking a method. Something which _would_ help solve the expression problem would be some form of pattern matching, but that is not on the table for JavaScript as far as I know.

**Comment by ivan-kleshnin** [comment]
> and for :: we already have Function.prototype.bind

Are you serious? Expression problem is solved by having dynamic this but it's a biggest secret of JS how it's coolest part is so ugly to write.

**Comment by aluanhaddad** [comment]
@ivan-kleshnin I agree about the power of `dynamic this` but the expression problem is usually thought of in terms of statically typed, class based languages. Since in JavaScript one can create ad hoc objects containing whatever structures and behaviors and and data one wishes, there's nothing which prevents you from extending a system with new behaviors and new structures. The hard thing is that since this is predicated on many axes of dynamism, statically reasoning about it and keeping it maintainable can be very difficult.

**Comment by ivan-kleshnin** [comment]
@aluanhaddad there are definitely multiple definitions of expression problem. Personally, I don't think that narrowing down such a sound definition for one particular case is right.

> there's nothing which prevents you from extending a system with new behaviors and new structures

By doing

```
Foo.prototype.whatever = whatever 
```

one makes a lot of assumptions. Enough to say it's not namespaced.

> The hard thing is that since this is predicated on many axes of dynamism, statically reasoning about it and keeping it maintainable can be very difficult.

Gradual typing is hard but I do think it's totally solvable in TypeScript give what they've done already.

**Comment by chicoxyzzy** [comment]
Notice this:
https://github.com/zenparsing/es-function-bind/issues/24#issuecomment-141102543

TL;DR
- it will not be `::`
- probably proposal will be splitted into separate "method extraction" and "functional pipelining" proposals
- it will not be added any soon

**Comment by ivan-kleshnin** [comment]
@chicoxyzzy thanks for the info, though bad news :(

**Comment by benjamingr** [comment]
Kevin needs a co champion https://github.com/zenparsing/es-function-bind/issues/24#issuecomment-225984655 - anyone from the TC care to step up?

**Comment by kitsonk** [comment]
@benjamingr a TypeScript issue is likely not the right forum to garner support for a TC39 proposal.  [ESDiscuss](https://esdiscuss.org/) is...

**Comment by benjamingr** [comment]
@kitsonk well, given Babel supports it and I'm considering migration to TypeScript, a champion from here would be a great help since it would mean early TypeScript support and easier migration.

**Comment by aluanhaddad** [comment]
@benjamingr it looks like whatever Babel supports will not come to be in its current form. Given that it is certainly not a reason to use Babel over TypeScript. It appears the team's conservatism on this proposal has been the right approach.

**Comment by zhirzh** [comment]
Is it possible to use `::` to bind data other than `this`?

As of July 2016, babel with stage-0 converts

```js
function X() {};
const x = new X();
function func(a, b) {}
x::func();
```

into 

```js
"use strict";

function X() {};
var x = new X();
function func(a, b) {}
func.call(x);
```

What I want is to do is to convert:

```js
"use strict";

function X() {};
const x = new X();
function func(a, b) {}
x::'paramA'::func(); // Or some other, more meaningful, syntax
```

into

```js
"use strict";

function X() {};
var x = new X();
function func(a, b) {}
func.bind(x, 'paramA').call();
```

**Comment by tjoskar** [comment]
@zhirzh, the syntax you suggests feels very strange and it will only works with strings? Right? Why don't you just: `x::func('paramA');`?
Anyway, this issue is not a place to discuss the syntax. Create a issue here:   https://github.com/zenparsing/es-function-bind or create a topic on [ESDiscuss](https://esdiscuss.org/).

**Comment by zhirzh** [comment]
It wasn't my intention to introduce new syntax. I was just asking if it is possible bind some data along with `this` - just like `function.bind(this, x, y);`

But with a litte google-fu, I found that `::` exists for 2 purposes: method extraction and virtual methods. That pretty much settles it.

**Comment by kube** [comment]
Is this planned to be added to a milestone soon ?

**Comment by SaschaNaz** [comment]
@kube No, it should be added on [ES proposal list](https://github.com/tc39/proposals) first.

**Comment by SaschaNaz** [comment]
@kube No, it should be added on [ES stage-1-and-above proposal list](https://github.com/tc39/proposals) first.

**Comment by aluanhaddad** [comment]
@SaschaNaz Indeed. Furthermore, I believe the proposal's original champion has left TC39.

**Comment by SaschaNaz** [comment]
@aluanhaddad Oh, that's unfortunate... Any more discussion should continue in [the right repo](https://github.com/tc39/proposal-bind-operator/issues/) anyway.

**Comment by mhegazy** [comment]
> Is this planned to be added to a milestone soon ?

As @aluanhaddad noted, the proposal seems to have stalled in TC39. Moreover, it is not clear what is the semantics of the new operator if it ever got implemented. There seems to be be two compelling use-cases, of using as sugar for `Function.prototype.bind`, and the other passing it by passing it as the first argument for the function.

From the TS perspective, we would like to get some consensus built in the committee first before approaching this to avoid breaking users in the future.

**Comment by SaschaNaz** [comment]
>There seems to be be two compelling use-cases, of using as sugar for Function.prototype.bind, and the other passing it by passing it as the first argument for the function.

Brian Terlson has [some idea to support both](https://github.com/tc39/proposal-bind-operator/issues/37):

>```
foo::bar(); // as normal
foo::bar(?); // approximately bar(foo)
foo::bar(1, ?); // approximately bar(1, foo)
```

**Comment by bterlson** [comment]
@mhegazy the story is even more complex - it's possible "extension" classes (eg. a way to declaratively re-open another class inside a particular scope) would solve many of these use cases as well.

I wouldn't say effort has stalled - I am still investigating and still want to see one or more of bind operator, pipeline, or extension classes in the language. But it will be some time before we get a good sense of the direction and so not implementing in TS for now seems like the best idea.

**Comment by aluanhaddad** [comment]
@bterlson thank you for the clarification and the update.

**Comment by benjamingr** [comment]
@bterlson if you're looking for historical context see https://esdiscuss.org/topic/scoped-binding-of-a-method-to-an-object

**Comment by AlexGalays** [comment]
I wish efforts were concentrated on the pipeline operator instead. Similar goal without all the idiosyncrasies of `this`. 

But I also agree there is no need to spread the TS team efforts with things that are not even in the standard. Could that kind of extensions be added as macros in user land perhaps?

**Comment by brunolm** [comment]
I like to use `::` operator on react files:

```typescript
<button onClick={ ::this.click } />
```

Instead of

```typescript
<button onClick={ this.click.bind(this) } />
```

But that doesn't seem to work on TypeScript yet

**Comment by aluanhaddad** [comment]
@brunolm given that the `::` operator has not had much, if any discussion, in recent TC39 meetings, and given that when it was discussed there was broad disagreement as to how it should work, the code you are writing is highly likely to be forever dependent on a very specific version of a very specific Babel plugin.
I would advise against using it altogether.

**Comment by AlexGalays** [comment]
@aluanhaddad  Yep. Also, bind() has pretty bad performances and it's better to bind once during class instantiation rather than every single time, on render().

**Comment by RyanCavanaugh** [comment]
Just a checkin, we are waiting for this proposal to reach stage 2. As usual please bribe your nearest TC39 rep if interested!

**Comment by kube** [comment]
I really hope that this operator will one day be treated as well, everybody seem to forget it, but it's really powerful.

https://github.com/mindeavor/es-pipeline-operator

**Comment by brunolm** [comment]
I wanted this operator to use on react only, but with the performance thing mentioned I lost interest in this operator. aluanhaddad's code looks good enough.

Thanks everyone for all the information I learned quite a bit :)

**Comment by Jamesernator** [comment]
@brunolm I've found the whole `.bind` thing to mostly be a fiction, now while it's true that `.bind` functions are slightly slower than regular functions because you need a new function object each time you bind it. In practice unless you're creating them en masse I don't think it'll impact you.

Honestly you only need to optimize where you can actually observe a difference! Randomly trying to optimize away features that was probably never the worst part of your performance is a great way to absolutely waste your time on nothing.

---

Also the bind operator will probably be better for optimizing than `.bind` anyway! The reason for that is that `2::someMethod(...someArgs)` could be optimized into `someMethod.apply(2, someArgs)`, like here's the performance of plain/method.bind/method.bind in Node.js:

Like take this for example:

```js


function* mapBind(iteratee) {
    for (const item of this) {
        yield iteratee(item)
    }
}

function* mapPlain(iterable, iteratee) {
    for (const item of iterable) {
        yield iteratee(item)
    }
}

const SIZE = 1000000

const randomData = new Array(SIZE)

for (let i = 0 ; i < SIZE ; i++) {
    randomData[i] = Math.random()
}

console.time('mapBind')
const bound = mapBind.bind(randomData)
const mapped1 = bound(x => x**2)
console.timeEnd('mapBind')

console.time('mapPlain')
const mapped2 = mapPlain(randomData, x => x**2)
console.timeEnd('mapPlain')

console.time('mapApply')
const mapped3 = mapBind.apply(randomData, [x => x**2])
console.timeEnd('mapApply')

console.time('mapBindRun')
const dat = [...mapped1]
console.timeEnd('mapBindRun')

console.time('mapPlainRun')
const dat2 = [...mapped2]
console.timeEnd('mapPlainRun')

console.time('mapApplyRun')
const dat3 = [...mapped3]
console.timeEnd('mapApplyRun')
```

Now using that I got:

```
mapBind: 0.327ms
mapPlain: 0.090ms
mapApply: 0.016ms
mapBindRun: 431.815ms
mapPlainRun: 298.362ms
mapApplyRun: 306.369ms
```

So yes it does appear true that `.bind` is slower however a good engine will optimize the binary form of `::` into `.apply` anyway so please don't dismiss features over nothing if they'd be legitimately useful to you, more support behind an idea means it's more likely to be prioritized for optimization!

In fact in FireFox they're already approximately equal so it's not even something fundamental to `.bind` itself:

```
// Times from FireFox
mapBind: timer started
mapBind: 1ms
mapPlain: timer started
mapPlain: 0.29ms
mapApply: timer started
mapApply: 0.3ms
mapBindRun: timer started
mapBindRun: 367.22ms
mapPlainRun: timer started
mapPlainRun: 369.41ms
mapApplyRun: timer started
mapApplyRun: 359.48ms
```

**Comment by graingert** [comment]
@Jamesernator fyi babel replaces `2::someMethod(...someArgs)` with `(_context = 2, someMethod).call(_context, ...someArgs);`


https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=stage-0&targets=Node-4&browsers=&builtIns=false&debug=false&code_lz=EwLhGcHsFsFMFlYBcAWkAmAKAdLqcBBAJwHNwBKIA&experimental=false&loose=false&spec=false&playground=true

**Comment by bloadvenro** [comment]
@aluanhaddad
 
Sorry, but I don't understand one thing:

For es5 compilation target

```
export default class {
  click = (event) => {
    //...
  };
}
```
gives

```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var default_1 = (function () {
    function default_1() {
        this.click = function (event) {
            // THIS FUNCTION WILL LOSE THIS CONTEXT <<<<<<<
        };
    }
    return default_1;
}());
exports.default = default_1;
```

**Comment by graingert** [comment]
@bloadvenro yeah arrows have lexical this context

**Comment by bloadvenro** [comment]
@graingert So. if I understand correctly, for es5 we still need ctor binding or in-place binding like `this.doSome.bind(this)`, because TS transpiles fat arrow functions into general ones.

**Comment by Yogu** [comment]
@bloadvenro No, you just didn't use `this` anywhere, so it is not bound.

```` 
export default class {
    private a = "test";  
  click = (event) => {
      console.log(this.a);
  };
}
```

transpiles to

```
define(["require", "exports"], function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var default_1 = (function () {
        function default_1() {
            var _this = this;
            this.a = "test";
            this.click = function (event) {
                console.log(_this.a);
            };
        }
        return default_1;
    }());
    exports.default = default_1;
});
```

**Comment by bloadvenro** [comment]
@Yogu Aaah, now I see, thank you!

**Comment by cvsguimaraes** [comment]
Apparently this is the only issue mentioning the pipeline operator. Could we have a thread for that operator to keep track of its status on TS?!

**Comment by graingert** [comment]
@cvsguimaraes https://github.com/Microsoft/TypeScript/issues/17718

**Comment by mhegazy** [comment]
Closing this issue. The `::` operator proposal is not active any longer, https://github.com/Microsoft/TypeScript/issues/17718 tracks the pipeline `|>` operator proposal.

**Comment by sHooKDT** [comment]
@mhegazy according to [tc39](https://github.com/tc39/proposals/blob/master/stage-0-proposals.md) it is active and still stage-0.
Why is it closed?
This issue prevents transition to TS for me (operator is widely presented in the codebase) and I would like to reopen it.

**Comment by graingert** [comment]
@sHooKDT you should run Babel with only the bind operator plugin enabled on your source, and save the output over the top

**Comment by sHooKDT** [comment]
> 
> 
> @sHooKDT you should run Babel with only the bind operator plugin enabled on your source, and save the output over the top

Yes, I tried today, but I'm not sure how to do this. Without other plugins like decorators babel is unable to parse my sources. I would be glad if you help me ;)

Anyway, the main point is not my transition, but the proposal that is still active. I see no reasons to close this now.

**Comment by DanielRosenwasser** [comment]
We'll revisit this if the proposal proceeds to stage 1.

**Comment by DanielRosenwasser** [comment]
Since locking this, https://github.com/tc39/proposal-extensions has reached stage 1. We still will not ship anything like this at stage 1, but I'll try to keep my word and reopen this issue for discussion.

**If all we get are comments about when we're planning to ship the feature, I will lock the issue again**, so please keep the conversation constructive

**Comment by RyanCavanaugh** [comment]
Closing since there's nothing to do until TC39 gets this to Stage 3
